Proposed structure for integration of benchmarks with pytest-benchmark

johannahaffner opened the pull request:
This PR is a proposal for what integration of benchmarks could look like. 

I suggest: 

- keeping a separate benchmark tests 
- running these with the usual familiar testing machinery
- keeping them out of the normal CI, at least once the benchmark collection is large enough to make inclusion impractical
- including extra information, such as the number of steps, the result of a nonlinear solve, and so on, which is useful for comparison to other implementations
- benchmarking results can be saved to a JSON file, which enables comparison across versions (check if we still get the same accuracy, if we now need more steps, and so on)

I have not included any compile time benchmarking yet, but I think this thing here can happily live next to differently flavoured pytest-codspeed benchmarks, such as discussed in https://github.com/patrick-kidger/equinox/issues/1001

I'm parsing a collection of CUTEST problems for now, and once you get rid of all the FORTRAN stuff and weird formatting these don't take up that much space and can either live here, or perhaps elsewhere if that is more practical.

patrick-kidger commented on tests/conftest.py:
```diff
@@ -11,3 +11,24 @@
     return eqxi.GetKey()
+
+
+# TODO: this can probably be done in a more streamlined fashion!
+# Skip benchmarks on default test runs. Request these explicitly (e.g. when stress-
+# testing new features).
```
It looks like this can be done with a `skipif`: https://stackoverflow.com/a/43938191/22545467

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,43 @@
+# Running benchmarks to test the performance of our solvers, using the pytest
+# benchmarking extension.
+#
+# To run the benchmarks, simply run `pytest benchmarks` from the root directory.
+# If you'd like to save the results, the flag `--benchmark-save=<file_path>` can be used
+# to save a .json file with stats and additional custom metrics. (Useful to compare
+# performance across versions.)
```
Let's put this in `benchmarks/README.md`

patrick-kidger commented on benchmarks/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+pytest
+pytest-benchmarks
```
Let's add this line to the `dev` group in `pyproject.toml`. (Just added in #130, #131) and remove the need for a separate requirements file here.

patrick-kidger reviewed (commented):
I think the basic structure of this looks good to me! I like being able to integrate this into pytest.

johannahaffner commented:
`skipif` works, but then the benchmark folder has to be in `tests`, since otherwise we cannot run `pytest benchmarks` separately. In that case it won't look in the `tests` folder, where the extra option is configured. I think it's fine to have the benchmarks folder be a sub-directory of `tests`, so I moved it. 

I also added an abstract base class for benchmark problems, and implemented this for our very favourite Rosenbrock function.

patrick-kidger commented on CONTRIBUTING.md:
````diff
@@ -38,6 +38,15 @@ pip install -e '.[dev]'
+benchmark tests with 
+
+```bash
+pytest --extensive
````
Maybe `--benchmark`?

johannahaffner replied:
I think there is to much confusion with the built-in flags then? We already have `--benchmark-save`, which is built-in and specifies where to save the results in benchmark format, and we also use the built-in benchmarking fixture with `@pytest.mark.benchmark`. 

I think that means that our own decorator and flags should be different! Maybe optx_benchmarks?

johannahaffner replied:
Or `@cutest` and `--cutest`? This will work fine until we integrate other benchmarking suites.

patrick-kidger replied:
You're anticipating a need to separate out solver benchmarks (cutest) vs compile-time benchmarks?

Anyway given that this is now in a separate directory, I'm realising that I think these will be ignored by pytest by default, and only ran if you're in the directory or if you run `pytest benchmarks` from the top-level, so perhaps we don't need a flag here at all?

johannahaffner replied:
No, they're not ignored by default, even if they are in a separate directory. 

They would be ignored if the names did not start with `test_`, but then we'd have the reverse problem.

I think runtime and compile time benchmarks can be run with the same problem / solver setup, but the `benchmark` fixture stores all the results and I think things might get muddled there? But happy to try.

johannahaffner replied:
I took a quick look at the structure of the `.json` output, and it seems wiser to have two different test functions for compile and runtime. Keeps things neater.

johannahaffner commented:
I'm now going with `@cutest` and benchmarks now have their own configuration. Both might be useful, since we might want to run different benchmarking suites (once we get there!) independently of one another, if they are geared toward different problems and require different solvers.

And we might want to tweak the configuration - to check how well we do in 32-bit, for instance - without affecting the main test suite.

patrick-kidger commented on benchmarks/cutest/problem.py:
```diff
@@ -0,0 +1,62 @@
+from jaxtyping import ArrayLike, PyTree, Scalar
+
+
+_Out = Union[Scalar, PyTree[ArrayLike]]
```
Nit, as we're now Py3.10+ then I'm using `Scalar | PyTree[ArrayLike]` now.

([pyupgrade](https://github.com/asottile/pyupgrade) is a good tool for doing these kinds of conversions automatically.)

johannahaffner replied:
Indeed, I have a visual preference for that too. Should we do that everywhere to be consistent?

patrick-kidger replied:
Ah, looks like I'm telling you porkies and that I haven't actually upgraded Optimistix in this way yet (I have started doing so on some of my other libraries.)

I think anything is fine then -- happy to run pyupgrade before or after this PR (but not in this PR, to avoid making it about two things), whicheer is easiest for you.

johannahaffner replied:
Haha alright! Let's postpone that for a bit then.

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,40 @@
+cutest = pytest.mark.skipif("not config.getoption('cutest')")
+
+
+def block_tree_until_ready(x):
+    dynamic, static = eqx.partition(x, eqx.is_inexact_array)
+    dynamic = jtu.tree_map(lambda x: x.block_until_ready(), dynamic)
+    return eqx.combine(dynamic, static)
```
FWIW the partition/combine operations will add a small but measurable amount of overhead, as this happens outside JIT.

I'd be tempted to suggest instead that outputs should be required to always only be JAX arrays, and to then call `jax.block_until_ready`.

johannahaffner replied:
We could return the value of the objective function at the solution and use that to determine if the problem has been solved to some degree of accuracy.

This is all that is available for 90+ % of CUTEST problems anyway.

More detailed metrics such as the number of iterations could be analysed separately, but we'd lose access to them here.

johannahaffner replied:
Ah you mean only returning the dynamic part of the solution? I understood this as doing something like `return solution.state.f_info.f` or something!

johannahaffner replied:
So we need at least `solution.RESULTS` which does not support `block_until_ready`. To accommodate this, I wrote a minimal wrapper that calls the compiled function, calls `block_until_ready` on the objective value and returns some extra stuff as well. 

Since the function does not actually do anything else, the overhead should be minimal. I don't know how it compares to a tree-mapped `block_until_ready`, though. If we really want to trim the microseconds then we could use `success = solution.result == optx.RESULTS.successful`, which is an array. 

I suggest we do that when benchmarking for publication, but otherwise it seems overkill - the usage of the benchmarking set is to catch regressions and test performance of new features, and in this case I think it is nice to have a .json file from which we can read informative results, not just minimal timing results.

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,40 @@
+@pytest.mark.parametrize("problem", cutest_unconstrained_problems)
+@pytest.mark.parametrize("minimiser", unconstrained_minimisers)
+def test_runtime_unconstrained_minimisers(benchmark, minimiser, problem):
+    compiled = eqx.filter_jit(
+        eqx.Partial(
+            optx.minimise, problem.objective, minimiser, problem.y0(), problem.args()
+        )
+    )
```
Likewise for benchmarking I think we should set up things using only `jax.jit`. We've done a pretty good job of optimising `eqx.filter_jit`, but using it is still explicitly an acknowledgement of being will to train a few microseconds of performance in favour of a simpler API.

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,40 @@
+    )
+
+    def wrapped():
+        return block_tree_until_ready(compiled())  # Returns an optx.Solution
```
Given that this consumes no arguments, then at least in principle it may be possible for XLA to constant-propagate at compile time all the way to the solution.

In practice I think it will only do that somewhat, but best to avoid this case I think.

johannahaffner replied:
You mean that all but the first run are going to be no-ops? Do I have to make it take array arguments to avoid that?

patrick-kidger replied:
At least in principle they could be.

In practice I think XLA is a bit milquetoast about actually doing constant propagation to that degree, but I could definitely still see it changing results.

johannahaffner replied:
Got it!

pfackeldey commented:
Hi @johannahaffner,
not sure if these thoughts are useful for you here, but:
- I think Github actions workers are not necessarily very homogeneous afaik. So it might be that you get different runtime values if you happen to land on a different worker. Of course, it's always better to have a server that you can use with a reliable small uncertainty on the runtime.
- you might want to restrict JAX to use a single thread only. I ran some benchmarks in the past and found it hard to restrict this (maybe it's better by now?). At least on linux you can use `taskset -c 0 pytest --cutest` to enforce this.

johannahaffner commented:
They are! Thank you :) 

> Github actions workers are not necessarily very homogeneous 

I was completely unaware of this! But that makes sense. I think depending on the size this benchmark set grows to (fingers crossed), we might not want to integrate it into the CI, but actually run it locally and add some scripts to help analyse the results. So this is another point in favour of maybe not doing this as part of our CI. 

> you might want to restrict JAX to use a single thread only.
And good point! We should take this into account when comparing results to other implementations.

johannahaffner commented:
I'm introducing a benchmark-time dependency on [sif2jax](https://github.com/johannahaffner/sif2jax) here, which is a current WIP of mine aiming to (have LLMs) port the complete CUTEst collection to clean, human-readable and functionally pure JAX implementations. 

I'll mark this as ready for review again when we have done further verification of the benchmark problems; currently about 247 benchmark problems pass tests against the Fortran implementation.

johannahaffner commented:
> * you might want to restrict JAX to use a single thread only. I ran some benchmarks in the past and found it hard to restrict this (maybe it's better by now?). At least on linux you can use `taskset -c 0 pytest --cutest` as a "solution".

This does seem to have improved? I've left this in as a commented-out suggestion, but I'm not quite sure what specific behaviour it is meant to prevent if enabled, @pfackeldey.

johannahaffner commented:
So I tested this a bit more rigorously and it seems like most things are working on our end! 

What we have right now:

- the ability to run benchmarks on  CUTEst problems provided by `sif2jax`, this introduces a benchmark-time dependency on a project of mine
- selectively run on problems up to a certain size
- compare against scipy for specific, defined solvers (other implementations, e.g. in JAXopt, optax, tensor flow can be added following the same pattern)
- analyse benchmark results - all the pytest-benchmark features, e.g. for comparison across runs, saved results with branch, commit info, time stamp, worktree status
- custom analysis: performance profiles of select solvers

I updated the README to reflect this.

Things that require a bit of thought still: 

- how to specify the solvers we're comparing against - solver tolerances do not necessarily have the same meaning in different libraries
- what additional metadata to store and to assess to check if solver comparisons are truly fair
- things like maybe allowing solver-specific maximum numbers of steps
- how/if to put benchmark results under version control so contributors can compare against each other's runs (pytest-benchmark saves these in folders that are specific to the platform (Darwin, Linux, ...), the Python implementation (CPython, ...), the Python version and the precision used, so no risk of mixing these up). Currently the benchmarking results are git ignored.
- I think it would be nice to ask Contributors that add a new solver if they could specify alternative implementations to benchmark against, since they are probably best positioned to judge the finer details of this -> this can go into CONTRIBUTING.md. Or does that make things sound too scary?

I'll mark this as ready to review for now :)

patrick-kidger commented on CONTRIBUTING.md:
````diff
@@ -38,6 +38,15 @@ pip install -e '.[dev]'
 ```
 
+If your changes could affect solver (or compilation) performance, please run the
+benchmark tests with 
````
+mention that they could run benchmarks before/after their change to test it.

Honestly I am starting to warm to the idea of using a service like codspeed etc to log benchmark results over time.

patrick-kidger commented on benchmarks/README.md:
````diff
@@ -0,0 +1,46 @@
+We're using [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/latest/) to assess performance. Benchmarks can be run with 
+
+```
+pytest benchmarks/ --cutest
````
Ah, looks like this needs syncing with `CONTRIBUTING.md`.

What happens if you only run `pytest benchmarks`?

What stops the benchmarks from being ran on just `pytest`?

How do you want to handle forward compatibility with any of our own benchmarks in the future? I'm anticipating that the `--cutest` flag would then no longer be the right name as we probably do want people to run our own benchmarks at the same time, and not make the distinction.

patrick-kidger commented on benchmarks/README.md:
```diff
@@ -0,0 +1,46 @@
+
+This will just run benchmarks on our own solvers - you can use saved results to compare performance to previous versions or different commits. Comparing against the last saved run is enabled with `pytest --benchmark-compare`, but specific iDs of previous runs may also be specified. (Consult the [documentation](https://pytest-benchmark.readthedocs.io/en/latest/) for more options.)
+
+For reproducibility, make sure that all your changes have been committed and your working tree is clean before running the benchmarks. `pytest-benchmark` will otherwise mark your benchmarking results as `dirty`. 
```
Does this matter? It's all local anyway.

patrick-kidger commented on benchmarks/README.md:
```diff
@@ -0,0 +1,46 @@
+For reproducibility, make sure that all your changes have been committed and your working tree is clean before running the benchmarks. `pytest-benchmark` will otherwise mark your benchmarking results as `dirty`. 
+Saved results include the commit, branch, version, and an exact timestamp by default.
+
+Note that benchmarks are run with `throw=False` enabled, since otherwise no result is written in the json file, but we do want to know if we failed to solve a problem.
```
If you wanted we could/should set `EQX_ON_ERROR=nan` here too. This is essentially equivalent to the above and is also a performance improvement.

(This has to happen before Equinox is imported.)

jpbrodrick89 replied:
Note that this is essential for vjp as throw=False does nothing there I believe

johannahaffner replied:
Argh I see that I might have boxed myself a little into a corner here - I could simplify the workflow and put the benchmarks back in the tests folder. This way we only get one `conftest` and one setup to deal with, and there are no stupid foot guns such as custom flags only being available if we're also specifying the folder name: e.g. 

`pytest benchmarks/ --benchmark-only--max-dimension=10` 

would work, but 

`pytest --benchmark-only --max-dimension=10` 

would not, which I had not previously realised.

But this also means that the environment variable applies to the whole test session, right? And while we want it for benchmarking, we don't want it for testing. 

Can I force a fresh environment for a specific testing module? Alternatively it might make sense to modify `--benchmark-only` so that if present, we do set the environment variable and not otherwise.

jpbrodrick89 replied:
Yes I would use a new environment for benchmarks (I find this hepful in my workflow as tests run faster than benchmarks and I can fail fast if something is wrong). Then no need to fiddle around with environment variables in `tests/` or `benchmarks/` and they are just set in the workflow actions (or by user). This is also helpful as it allows the user to run with different setting to compare impact.

johannahaffner replied:
In Github workflows? That makes sense! Locally we default to no benchmark runs, and I just discovered that I can monkeypatch the environment variables for specific tests only - would that work, @patrick-kidger?

patrick-kidger commented on benchmarks/README.md:
````diff
@@ -0,0 +1,46 @@
+The usage of this is 
+
+```bash
+python3 benchmarks/profile.py solver1 solver2 platform python_version precision iD
````
`python3` -> `python`

patrick-kidger commented on benchmarks/conftest.py:
```diff
@@ -0,0 +1,43 @@
+jax.config.update("jax_enable_x64", True)
+jax.config.update("jax_numpy_rank_promotion", "raise")
+jax.config.update("jax_numpy_dtype_promotion", "standard")
+# Remark Peter: limit JAX to single-thread with taskset -c 0 pytest --cutest
```
Note that `taskset` is Linux-only, I believe.

patrick-kidger commented on benchmarks/profile.py:
```diff
@@ -0,0 +1,228 @@
+import jax.numpy as jnp
+import matplotlib.pyplot as plt  # pyright: ignore[reportMissingImports]
+import numpy as np  # pyright: ignore[reportMissingImports]
+import pandas as pd  # pyright: ignore[reportMissingImports]
```
Is pandas that useful here? (It seems fairly minimal below.)

To me this is in the same bucket as libraries like huggingface and lightning - reasonable for hacking around but not a thing I'd like to check in. (I realise this is a somewhat unusual position.)

patrick-kidger commented on benchmarks/profile.py:
```diff
@@ -0,0 +1,228 @@
+
+def find_benchmark_run(benchmarks_dir: pathlib.Path, args: list) -> dict:
+    """Find the benchmark run data based on the provided arguments."""
+    platform, python_version, precision, iD = args[-4:]
```
Maybe we should use https://github.com/google/python-fire to make the argument-parsing nice?

I think it's okay to introduce the extra dependency as we can put it in the dev dependency list.

patrick-kidger commented on pyproject.toml:
```diff
@@ -46,7 +46,9 @@ dev = [
-  "ruff>=0.11.6"
+  "pytest-benchmark>=5.1.0",
+  "ruff>=0.11.6",
+  "sif2jax>=0.0.3"
```
I think this needs extending with pandas (if we keep it) and matplotlib.

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,132 @@
+        )
+    )
+
+    _ = solve(problem.y0())  # Warm up
```
I don't think we need the `_ =`

jpbrodrick89 replied:
Any reason not to be more explicit with what you're doing and just do `.lower(...).compile()`?

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,132 @@
+
+    def wrapped(y0):
+        solution = solve(y0)
+        objective_value = solution.state.f_info.f.block_until_ready()
```
Maybe just `jax.block_until_ready(solution)`?

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,132 @@
+        return objective_value, solution.result, num_steps
+
+    # Benchmark the runtime of the compiled function
+    values = benchmark.pedantic(wrapped, args=(problem.y0(),), rounds=5, iterations=1)
```
How does the benchmarking work here btw? As compared to e.g. `min(timeit.repeat(..., number=..., repeat=...))` or the ipython `%%timeit`? (E.g. the former uses minimisation to handle one-sided noise, whilst the latter automatically figures out how many runs to perform.)

patrick-kidger reviewed (commented):
I really like the look of this.

jpbrodrick89 commented:
Hi guys, I will chime in a tiny bit more detail this afternoon as I've been trying a lot of this on one of my repos. Unfortunately CodSpeeds usual instrumentation provides no value for Jax functions as everything counts as a system calls which is not trackable by valgrind. Therefore to use their tooling you have to use their wall runner where costs can mount up pretty quickly. I had 360 micro benchmarks and used up the free 2 hours in just four commits. Also, even on their bare metal wall runner there is still some noise probably due the time of actually retrieving the function from the jax cache rather than running the actual floap. Trying to work out if I can make this more useable next week. At least they've updated their docs in the last couple weeks quite substantially.
@johannahaffner happy to have a call if helpful.

jpbrodrick89 commented on .github/workflows/codspeed.yml:
```diff
@@ -0,0 +1,40 @@
+
+#       - name: Run benchmarks
+#         uses: CodSpeedHQ/action@v3
+#         with:
```
This is where you can add the env variables
```suggestion
#         with:
#           env:
#             EQX_ON_ERROR: "nan"
```

pfackeldey commented:
> I opted against using taskset, because I don't really understand what it would/should do

Fyi, [`taskset`](https://www.man7.org/linux/man-pages/man1/taskset.1.html) lets you restrict the number of cores seen by a process. E.g. `taskset -c 1 python jax_code.py` would _only_ use 1 core and not however many are available on the machine. IIRC jax is rather aggressive about this and uses whatever is available.

I assume this only makes sense if you are interested in having a history of benchmarks over multiple CI runs, _and_ given the premise that Github Actions runner may have different number of cores. (I'm not sure how homogeneous GitHub actions are?).

johannahaffner commented:
> I assume this only makes sense if you are interested in having a history of benchmarks over multiple CI runs, _and_ given the premise that Github Actions runner may have different number of cores. (I'm not sure how homogeneous GitHub actions are?).
> 
> (This only works on linux, and if the cores differ in clock frequency between different workers, that also doesn't help of course)

Thank you for the explanation! Does this mean that you can get around some of the variability in CI runs? Apparently how fast things are in there depends on a lot of factors. That is actually what codspeed advertises - they claim that they get the noise down from around 30 % to 1 %; and they do that through some clever isolation of the bit to be benchmarked, which is then run just once. But this does, apparently, not really mesh well with JAX. 

FWIW I've come to the conclusion that the pragmatic way forward is to just use the benchmarks locally during development.

jpbrodrick89 commented on tests/test_benchmarks.py:
```diff
@@ -0,0 +1,151 @@
+            throw=False,
+        )
+    )
+
```
```suggestion
y0 = jax.tree.map(lambda x: jax.device_put(x) if eqx.is_array(x) else x, problem.y0())
```
This makes sure that device transfer times are not includedi n benchmarks which can be significant for GPU's.

johannahaffner commented:
I think this is ready! Thanks to all of you for your help, @patrick-kidger @jpbrodrick89 and @pfackeldey. 

Small updates made since last synopsis:

- accepted @jpbrodrick89's suggestion to use device put
- added a comment to use task set in `run_tests.yml` - for future reference, once this becomes part of the CI
- bumped the minimum version of `sif2jax` to one that is compatible with strict dtype and rank promotion rules
- moved the benchmarking script back to the `benchmarks` folder, with its own `conftest` - this is to make sure that we have a separate benchmarking environment, which now uses `EQX_ON_ERROR = "nan"` and does not monkeypatch anything.
- updated CONTRIBUTING.md to match.

jpbrodrick89 commented:
Exciting time @johannahaffner! 🚀 It looks like you set eqx on error as a global python variable rather than an os environment variable, see here for different options for setting env variables in pytest https://stackoverflow.com/questions/36141024/how-to-pass-environment-variables-to-pytest

The way I do is just to set on command line with `EQX_ON_ERROR=nan pytest....`

johannahaffner commented:
> Awesome! Not at the desk right now to to check, but i think you should be able to get rid of the noqa e402s now as os is meant to be ignored (https://docs.astral.sh/ruff/rules/module-import-not-at-top-of-file/) I'm not sure about the i001would be nice if this had a similar exception but it's not documented.

Indeed! And `I001` does not make exceptions.

johannahaffner commented:
I tried some benchmarking on our brand new L-BFGS and the current setup, using `jax.jit` exclusively, complains about the jaxpr in its state. I'll see if I can figure out a way around this that also works for compilation benchmarking.

johannahaffner commented:
@patrick-kidger what do you think about merging this as-is, with the `jax.jit`/ jaxpr in solver state blocker merely documented, but not resolved?

We do have support for runtime benchmarking, we just cannot benchmark the (AOT) compilation time for a subset of solvers. This PR is now a little old - we did think about it some three months ago, which is also when it was mostly done. To be honest, most things were no longer very fresh in my mind, and I had to re-run some benchmarks and generate some figures to convince myself that things are as they should be. 

Merging this with support for runtime benchmarking only seems to be a pragmatic way forward here.

patrick-kidger commented on docs/_static/quadratic_bowl.py:
```diff
@@ -14,19 +14,19 @@ def quadratic_bowl(x, y):
+ax.zaxis.set_rotate_label(False)  # pyright: ignore
+ax.xaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))  # pyright: ignore
+ax.yaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))  # pyright: ignore
+ax.zaxis.set_pane_color((1.0, 1.0, 1.0, 1.0))  # pyright: ignore
```
Nit: Can these be `pyright: ignore[somethingSpecific]`?

(I didn't used to do this, but I think it's a best practice.)

patrick-kidger reviewed (commented):
Okay nits only! This basically LGTM.

johannahaffner commented on CONTRIBUTING.md:
````diff
@@ -38,6 +38,13 @@ pip install -e '.[dev]'
+If your changes could affect solver (or compilation) performance, please run the benchmark tests with 
+
+```bash
+pytest benchmarks/ --benchmark-only
````
Almost - 

`--benchmark-only` without the `benchmarks/` would collect everything in `tests/` and then skip it. 

`benchmarks/` prevents the collection of everything in `tests/`, but it is only necessary for the definition of custom options in `benchmarks/conftest.py`, such as filtering by the minimum and maximum number of dimensions, or running `scipy` tests. If `benchmarks/` is not used, then passing one of our options will result in an error.

johannahaffner commented on CONTRIBUTING.md:
````diff
@@ -58,3 +65,40 @@ mkdocs serve
+You can find a script to analyse benchmark results in `benchmarks/profile.py`. Run it with
+
+```bash
+python benchmarks/profile.py platform python_version precision iD kind *solver_names
````
The Python version on which the tests are run is detected automatically and results are saved with the appropriate path. 

Detecting the version from which this script is ran to generate the plots would mean that we assume that these are always the same, which is probably true most of the time but not something I would like to bake in.

johannahaffner replied:
See reply in https://github.com/patrick-kidger/optimistix/pull/200.

johannahaffner commented on benchmarks/profile.py:
```diff
@@ -0,0 +1,306 @@
+                    "successful": is_successful,
+                }
+
+    return OrderedDict(solver_data)
```
I only got rid of a bug after switching to this, on Python 3.13.

Results were switched, and getting a correct comparison of our LBFGS vs. our BFGS and `optax.lbfgs` was only possible after switching to `OrderedDict`.  Not sure why - perhaps because of the `set` operations?

patrick-kidger replied:
Hmm, that seems very strange. I would have thought that this call basically does nothing.

Either way FWIW, to me this is suggestive of a code smell. I'd generally argue that it's best to treat dicts as unordered, as depending on their ordering is often a subtle point missed during refactors.

Staring at your code I don't seem an obvious issue though. The `set` operations would result in a random ordering within `get_relative_performance`, but I don't think that matters...

patrick-kidger commented on benchmarks/test_benchmarks.py:
```diff
@@ -0,0 +1,216 @@
+        )
+    )
+    y0 = jax.tree.map(
+        lambda x: jax.device_put(x) if eqx.is_array(x) else x, problem.y0()
```
Oh sorry, I'm misreading the lambda. Yup ignore me.
